anonymous_helpline_chatbot/core/invitations.py

```diff
@@ -1,89 +1,131 @@
-                return 1
-
-            # In a conversation already
-            if self.conversations_controller.get_conversing(tg_client_id) != ((None, None), (None, None)):
-                return 3
-
-            msg_ids = []
-            for tg_operator_id in free_operators:
-                try:
-                    msg_ids.append((
-                        tg_operator_id,
-                        self.send_invitation_callback(tg_operator_id, tg_client_id,
-                                                      f"Пользователь №{local_client_id} хочет побеседовать. Нажмите "
-                                                      "кнопку ниже, чтобы стать его оператором")
-                    ))
-                except telebot.apihelper.ApiException:
-                    print("Telegram API Exception while sending out operators invitations:", file=stderr)
-                    print(format_exc(), file=stderr)
-
-            self._operators_invitations_messages[tg_client_id] = msg_ids
-
-        return 0
-
-    def clear_invitation_messages(self, tg_client_id: int) -> bool:
+        with self._conn_pool.PrettyCursor() as cursor:
+            # Prevent any invitations from being sent or deleted by parallel transactions until this one completes,
+            # because, for example, a parallel transaction might try to delete an invitation which we are going to send,
+            # but have not sent yet
+            cursor.execute("LOCK TABLE sent_invitations IN SHARE MODE")
+
+            """
+            Explanation of the query below:
+
+            This query selects chat ids of all the operators which should receive an invitation to the given client
+            (i.e. they are currently not in conversations, and they haven't yet been sent an invitation to the client).
+            We select them with the two joins in the following way (of course, PostgreSQL doesn't do exactly what I say
+            here, it performs optimizations. But the result is exactly as if it was doing the following):
+            1. SELECT all the users, which are operators (see `WHERE users.is_operator` in the end of the query)
+            2. If the client himself is an operator, he shouldn't receive a notification. Remove him from the
+                resulting set (`WHERE ... AND users.chat_id != <client_chat_id>`)
+            3. LEFT OUTER JOIN the selected users with conversations (note the `ON` clause: the rows are considered
+                matching if the user is in a conversation in any role) and forget users which are in conversations
+                (`WHERE ... AND conversations.operator_chat_id IS NULL`)
+            4. Another unusual LEFT OUTER JOIN: we join the remaining users with `sent_invitations`, but the join is on
+                an interesting condition: the user row matches an invitation row if the user is the invitation operator
+                AND the invitation client is the client we're currently processing. This way we get rid of other
```

Comment:
Mm, an iNtErEsTiNg condition
```suggestion
            4. Another unusual LEFT OUTER JOIN: we join the remaining users with `sent_invitations`, where a user row
                is considered to be matching an invitation row if the user is the operator, to which the invitation
                was sent AND the invitation client is the client we're currently processing. This way we get rid of other
```